Add shellcheck to CI

For bash scripts in `scripts/dev` and for `misc/userscripts` it'd be nice to run [shellcheck](https://github.com/koalaman/shellcheck) on Travis.

Current output:

```
In misc/userscripts/open_download line 79:
    ls -Q --quoting-style escape    -h  -o -1 -A -t "${DOWNLOAD_DIR}/" \
    ^-- SC2010: Don't use ls | grep. Use a glob or a for loop with a condition to allow non-alphanumeric filenames.


In misc/userscripts/open_download line 97:
       | $ROFI_CMD "${rofi_default_args[@]}" $ROFI_ARGS) || true
                                             ^-- SC2086: Double quote to prevent globbing and word splitting.


In misc/userscripts/password_fill line 239:
                if $GPG "${GPG_OPTS}" -d "$passfile" \
                         ^-- SC2128: Expanding an array without an index only gives the first element.


In misc/userscripts/password_fill line 272:
        done < <($GPG "${GPG_OPTS}" -d "$path" )
                       ^-- SC2128: Expanding an array without an index only gives the first element.


In misc/userscripts/password_fill line 286:
            if [[ "$line" =~ "attribute.username = " ]] ; then
                             ^-- SC2076: Don't quote rhs of =~, it'll match literally rather than as a regex.


In misc/userscripts/password_fill line 306:
    source "$PWFILL_CONFIG"
    ^-- SC1090: Can't follow non-constant source. Use a directive to specify location.


In misc/userscripts/qutebrowser_viewsource line 29:
curl "$QUTE_URL" > $path
                   ^-- SC2086: Double quote to prevent globbing and word splitting.


In misc/userscripts/qutedmenu line 47:
    source "$optsfile"
    ^-- SC1090: Can't follow non-constant source. Use a directive to specify location.


In misc/userscripts/taskadd line 28:
msg="$(task add $title $@ 2>&1)"
                ^-- SC2086: Double quote to prevent globbing and word splitting.
                       ^-- SC2068: Double quote array expansions to avoid re-splitting elements.


In misc/userscripts/taskadd line 33:
    echo "message-info '$msg'" >> $QUTE_FIFO
                                  ^-- SC2086: Double quote to prevent globbing and word splitting.


In misc/userscripts/taskadd line 35:
    echo "message-error '$msg'" >> $QUTE_FIFO
                                   ^-- SC2086: Double quote to prevent globbing and word splitting.

In scripts/dev/quit_segfault_test.sh line 10:
    while (( $exit == 0)); do
             ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In scripts/dev/quit_segfault_test.sh line 11:
        duration=$(($RANDOM%10000))
                    ^-- SC2004: $/${} is unnecessary on arithmetic variables.

In scripts/dev/ci/travis_install.sh line 1:
# vim: ft=sh fileencoding=utf-8 sts=4 sw=4 et:
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.


In scripts/dev/ci/travis_install.sh line 29:
            echo -e "\n${ANSI_RED}The command \"$@\" failed. Retrying, $count of 3.${ANSI_RESET}\n" >&2
                                                ^-- SC2145: Argument mixes string and array. Use * or separate argument.


In scripts/dev/ci/travis_install.sh line 34:
        count=$(($count + 1))
                 ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In scripts/dev/ci/travis_install.sh line 39:
        echo -e "\n${ANSI_RED}The command \"$@\" failed 3 times.${ANSI_RESET}\n" >&2
                                            ^-- SC2145: Argument mixes string and array. Use * or separate argument.


In scripts/dev/ci/travis_install.sh line 57:
    travis_retry sudo -H python -m pip install -r misc/requirements/requirements-$1.txt
                                                                                 ^-- SC2086: Double quote to prevent globbing and word splitting.


In scripts/dev/ci/travis_install.sh line 114:
        apt_install xvfb $pyqt_pkgs libpython3.4-dev
                         ^-- SC2086: Double quote to prevent globbing and word splitting.


In scripts/dev/ci/travis_install.sh line 118:
        apt_install $pyqt_pkgs libpython3.4-dev
                    ^-- SC2086: Double quote to prevent globbing and word splitting.


In scripts/dev/ci/travis_install.sh line 125:
        apt_install $pyqt_pkgs asciidoc libpython3.4-dev
                    ^-- SC2086: Double quote to prevent globbing and word splitting.


In scripts/dev/ci/travis_run.sh line 4:
    docker run --privileged -v $PWD:/outside thecompiler/qutebrowser:$DOCKER
                               ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                                     ^-- SC2086: Double quote to prevent globbing and word splitting.


In scripts/dev/ci/travis_run.sh line 9:
    tox -e $TESTENV -- "${args[@]}"
           ^-- SC2086: Double quote to prevent globbing and word splitting.
```

Some non-issues in there, and some stuff which looks like a real problem.

Abandoned patch from #1697:

```patch
From 4aba13bbd9fe0f5a4be12af86c2cc46f7e1ce291 Mon Sep 17 00:00:00 2001
From: neeasade <nathanisom27@gmail.com>
Date: Thu, 28 Jul 2016 14:14:13 -0400
Subject: [PATCH 1/3] shell check fixes.

---
 misc/userscripts/open_download          |  3 ++-
 misc/userscripts/qutebrowser_viewsource | 11 +++++++++--
 misc/userscripts/qutedmenu              | 12 +++---------
 misc/userscripts/taskadd                |  7 +++----
 scripts/dev/ci/travis_install.sh        | 15 ++++++++-------
 scripts/dev/ci/travis_run.sh            |  6 +++---
 scripts/dev/quit_segfault_test.sh       | 10 ++++------
 7 files changed, 32 insertions(+), 32 deletions(-)

diff --git a/misc/userscripts/open_download b/misc/userscripts/open_download
index 54c9d3e971..532de611c0 100755
--- a/misc/userscripts/open_download
+++ b/misc/userscripts/open_download
@@ -76,6 +76,7 @@ crop-first-column() {
 ls-files() {
     # add the slash at the end of the download dir enforces to follow the
     # symlink, if the DOWNLOAD_DIR itself is a symlink
+    # shellcheck disable=SC2010
     ls -Q --quoting-style escape    -h  -o -1 -A -t "${DOWNLOAD_DIR}/" \
         | grep '^[-]' \
         | cut -d' ' -f3- \
@@ -94,7 +95,7 @@ fi
 line=$(printf "%s\n" "${entries[@]}"  \
        | crop-first-column 55 \
        | column -s $'\t' -t \
-       | $ROFI_CMD "${rofi_default_args[@]}" $ROFI_ARGS) || true
+       | $ROFI_CMD "${rofi_default_args[@]}" "$ROFI_ARGS") || true
 if [ -z "$line" ];  then
     exit 0
 fi
diff --git a/misc/userscripts/qutebrowser_viewsource b/misc/userscripts/qutebrowser_viewsource
index 535855bb3c..3ed7e01cac 100755
--- a/misc/userscripts/qutebrowser_viewsource
+++ b/misc/userscripts/qutebrowser_viewsource
@@ -24,9 +24,16 @@
 # Caveat: Does not use authentication of any kind. Add it in if you want it to.
 #
 
+TERMINAL=${TERMINAL:-urxvt}
+EDITOR=${EDITOR:-vim}
+
 path=/tmp/qutebrowser_$(mktemp XXXXXXXX).html
+curl "$QUTE_URL" > "$path"
 
-curl "$QUTE_URL" > $path
-urxvt -e vim "$path"
+unset _command
+[[ $EDITOR =~ emacs|gedit|leafpad|atom ]] && _command="$EDITOR $path"
+[ -z "$_command" ] && _command="$TERMINAL -e \"vim $path\""
 
+# debug: this would be eval not echo.
+echo "$_command"
 rm "$path"
diff --git a/misc/userscripts/qutedmenu b/misc/userscripts/qutedmenu
index 3f8b135141..b753c6d5ba 100755
--- a/misc/userscripts/qutedmenu
+++ b/misc/userscripts/qutedmenu
@@ -35,17 +35,11 @@ get_selection() {
 
 # Main
 # https://github.com/halfwit/dotfiles/blob/master/.config/dmenu/font
-if [[ -s $confdir/dmenu/font ]]; then
-    read -r font < "$confdir"/dmenu/font
-fi
+[[ -s $confdir/dmenu/font ]] && read -r font < "$confdir"/dmenu/font
 
-if [[ $font ]]; then
-    opts+=(-fn "$font")
-fi
+[[ $font ]] && opts+=(-fn "$font")
 
-if [[ -s $optsfile ]]; then
-    source "$optsfile"
-fi
+[[ -s $optsfile ]] && source "$optsfile"
 
 url=$(get_selection)
 url=${url/*http/http}
diff --git a/misc/userscripts/taskadd b/misc/userscripts/taskadd
index 6add71c68c..d4ecd4fe2e 100755
--- a/misc/userscripts/taskadd
+++ b/misc/userscripts/taskadd
@@ -25,12 +25,11 @@
 [[ $QUTE_MODE == 'hints' ]] && title=$QUTE_SELECTED_TEXT || title=$QUTE_TITLE
 
 # try to add the task and grab the output
-msg="$(task add $title $@ 2>&1)"
+msg="$(task add "$title" "$*" 2>&1)"
 
 if [[ $? == 0 ]]; then
     # annotate the new task with the url, send the output back to the browser
     task +LATEST annotate "$QUTE_URL"
-    echo "message-info '$msg'" >> $QUTE_FIFO
-else
-    echo "message-error '$msg'" >> $QUTE_FIFO
+    echo "message-info '$msg'" >> "$QUTE_FIFO"
+else echo "message-error '$msg'" >> "$QUTE_FIFO"
 fi
diff --git a/scripts/dev/ci/travis_install.sh b/scripts/dev/ci/travis_install.sh
index 435af63899..ec60621c5f 100644
--- a/scripts/dev/ci/travis_install.sh
+++ b/scripts/dev/ci/travis_install.sh
@@ -1,3 +1,4 @@
+#/usr/bin/env bash
 # vim: ft=sh fileencoding=utf-8 sts=4 sw=4 et:
 
 # Copyright 2016 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
@@ -26,17 +27,17 @@ travis_retry() {
     local count=1
     while (( count < 3 )); do
         if (( result != 0 )); then
-            echo -e "\n${ANSI_RED}The command \"$@\" failed. Retrying, $count of 3.${ANSI_RESET}\n" >&2
+            echo -e "\n${ANSI_RED}The command \"$*\" failed. Retrying, $count of 3.${ANSI_RESET}\n" >&2
         fi
         "$@"
         result=$?
         (( result == 0 )) && break
-        count=$(($count + 1))
+        count=$(( count + 1 ))
         sleep 1
     done
 
     if (( count > 3 )); then
-        echo -e "\n${ANSI_RED}The command \"$@\" failed 3 times.${ANSI_RESET}\n" >&2
+        echo -e "\n${ANSI_RED}The command \"$*\" failed 3 times.${ANSI_RESET}\n" >&2
     fi
 
     return $result
@@ -54,7 +55,7 @@ brew_install() {
 
 pip_install() {
     # this uses python2
-    travis_retry sudo -H python -m pip install -r misc/requirements/requirements-$1.txt
+    travis_retry sudo -H python -m pip install -r "misc/requirements/requirements-$1.txt"
 }
 
 npm_install() {
@@ -111,18 +112,18 @@ tox --version
 case $TESTENV in
     py34-cov)
         pip_install codecov
-        apt_install xvfb $pyqt_pkgs libpython3.4-dev
+        apt_install xvfb "$pyqt_pkgs" libpython3.4-dev
         check_pyqt
         ;;
     pylint|vulture)
-        apt_install $pyqt_pkgs libpython3.4-dev
+        apt_install "$pyqt_pkgs" libpython3.4-dev
         check_pyqt
         ;;
     flake8)
         apt_install libpython3.4-dev
         ;;
     docs)
-        apt_install $pyqt_pkgs asciidoc libpython3.4-dev
+        apt_install "$pyqt_pkgs" asciidoc libpython3.4-dev
         asciidoc --version
         check_pyqt
         ;;
diff --git a/scripts/dev/ci/travis_run.sh b/scripts/dev/ci/travis_run.sh
index 6df22da1e8..ebbd120246 100644
--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -1,10 +1,10 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 if [[ $DOCKER ]]; then
-    docker run --privileged -v $PWD:/outside thecompiler/qutebrowser:$DOCKER
+    docker run --privileged -v "$PWD:/outside thecompiler/qutebrowser:$DOCKER"
 else
     args=()
     [[ $TESTENV == docs ]] && args=('--no-authors')
 
-    tox -e $TESTENV -- "${args[@]}"
+    tox -e "$TESTENV" -- "${args[@]}"
 fi
diff --git a/scripts/dev/quit_segfault_test.sh b/scripts/dev/quit_segfault_test.sh
index 655eb262a2..90933a9650 100755
--- a/scripts/dev/quit_segfault_test.sh
+++ b/scripts/dev/quit_segfault_test.sh
@@ -1,14 +1,12 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
-if [[ $PWD == */scripts ]]; then
-    cd ..
-fi
+[[ $PWD == */scripts ]] && cd ..
 
 echo > crash.log
 while :; do
     exit=0
-    while (( $exit == 0)); do
-        duration=$(($RANDOM%10000))
+    while (( exit == 0)); do
+        duration=$(( RANDOM % 10000))
         python3 -m qutebrowser --debug ":later $duration quit" http://www.heise.de/
         exit=$?
     done

From 8141a18aacbdeb8e4ef950c6d724f8842ed83879 Mon Sep 17 00:00:00 2001
From: neeasade <nathanisom27@gmail.com>
Date: Thu, 28 Jul 2016 15:17:06 -0400
Subject: [PATCH 2/3] password_fill updates, and correction.

---
 misc/userscripts/password_fill | 7 ++++---
 misc/userscripts/taskadd       | 3 ++-
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/misc/userscripts/password_fill b/misc/userscripts/password_fill
index 5d709512ce..8085a97b90 100755
--- a/misc/userscripts/password_fill
+++ b/misc/userscripts/password_fill
@@ -236,7 +236,7 @@ pass_backend() {
         if ((match_line)) ; then
             # add entries with matching URL-tag
             while read -r -d "" passfile ; do
-                if $GPG "${GPG_OPTS}" -d "$passfile" \
+                if $GPG "${GPG_OPTS[@]}" -d "$passfile" \
                      | grep --max-count=1 -iE "${match_line_pattern}${url}" > /dev/null
                 then
                     passfile="${passfile#$PREFIX}"
@@ -269,7 +269,7 @@ pass_backend() {
                     break
                 fi
             fi
-        done < <($GPG "${GPG_OPTS}" -d "$path" )
+        done < <($GPG "${GPG_OPTS[@]}" -d "$path" )
     }
 }
 # =======================================================
@@ -283,7 +283,8 @@ secret_backend() {
     query_entries() {
         local domain="$1"
         while read -r line ; do
-            if [[ "$line" =~ "attribute.username = " ]] ; then
+            # shellcheck disable=SC1001
+            if [[ "$line" =~ attribute.username\ \= ]] ; then
                 files+=("$domain ${line#${BASH_REMATCH[0]}}")
             fi
         done < <( secret-tool search --unlock --all domain "$domain" 2>&1 )
diff --git a/misc/userscripts/taskadd b/misc/userscripts/taskadd
index d4ecd4fe2e..9c70fb9785 100755
--- a/misc/userscripts/taskadd
+++ b/misc/userscripts/taskadd
@@ -31,5 +31,6 @@ if [[ $? == 0 ]]; then
     # annotate the new task with the url, send the output back to the browser
     task +LATEST annotate "$QUTE_URL"
     echo "message-info '$msg'" >> "$QUTE_FIFO"
-else echo "message-error '$msg'" >> "$QUTE_FIFO"
+else
+    echo "message-error '$msg'" >> "$QUTE_FIFO"
 fi

From 4e2a8c0f0b2b046a40066ac2ba9a7e8f5d6877bb Mon Sep 17 00:00:00 2001
From: Nathan Isom <nathanisom27@gmail.com>
Date: Fri, 29 Jul 2016 07:01:46 -0500
Subject: [PATCH 3/3] style and feedback changes to shell scripts.

---
 misc/userscripts/password_fill          |  3 +--
 misc/userscripts/qutebrowser_viewsource |  3 +--
 scripts/dev/ci/travis_install.sh        | 11 +++++++----
 scripts/dev/ci/travis_run.sh            |  2 +-
 scripts/dev/quit_segfault_test.sh       |  4 ++--
 5 files changed, 12 insertions(+), 11 deletions(-)

diff --git a/misc/userscripts/password_fill b/misc/userscripts/password_fill
index 8085a97b90..8c3ab06236 100755
--- a/misc/userscripts/password_fill
+++ b/misc/userscripts/password_fill
@@ -283,8 +283,7 @@ secret_backend() {
     query_entries() {
         local domain="$1"
         while read -r line ; do
-            # shellcheck disable=SC1001
-            if [[ "$line" =~ attribute.username\ \= ]] ; then
+            if [[ "$line" == "attribute.username ="* ]] ; then
                 files+=("$domain ${line#${BASH_REMATCH[0]}}")
             fi
         done < <( secret-tool search --unlock --all domain "$domain" 2>&1 )
diff --git a/misc/userscripts/qutebrowser_viewsource b/misc/userscripts/qutebrowser_viewsource
index 3ed7e01cac..c5cce842f3 100755
--- a/misc/userscripts/qutebrowser_viewsource
+++ b/misc/userscripts/qutebrowser_viewsource
@@ -34,6 +34,5 @@ unset _command
 [[ $EDITOR =~ emacs|gedit|leafpad|atom ]] && _command="$EDITOR $path"
 [ -z "$_command" ] && _command="$TERMINAL -e \"vim $path\""
 
-# debug: this would be eval not echo.
-echo "$_command"
+eval "$_command"
 rm "$path"
diff --git a/scripts/dev/ci/travis_install.sh b/scripts/dev/ci/travis_install.sh
index ec60621c5f..f14c24450d 100644
--- a/scripts/dev/ci/travis_install.sh
+++ b/scripts/dev/ci/travis_install.sh
@@ -1,4 +1,4 @@
-#/usr/bin/env bash
+#!/usr/bin/env bash
 # vim: ft=sh fileencoding=utf-8 sts=4 sw=4 et:
 
 # Copyright 2016 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
@@ -112,18 +112,21 @@ tox --version
 case $TESTENV in
     py34-cov)
         pip_install codecov
-        apt_install xvfb "$pyqt_pkgs" libpython3.4-dev
+        # shellcheck disable=SC2086
+        apt_install xvfb $pyqt_pkgs libpython3.4-dev
         check_pyqt
         ;;
     pylint|vulture)
-        apt_install "$pyqt_pkgs" libpython3.4-dev
+        # shellcheck disable=SC2086
+        apt_install $pyqt_pkgs libpython3.4-dev
         check_pyqt
         ;;
     flake8)
         apt_install libpython3.4-dev
         ;;
     docs)
-        apt_install "$pyqt_pkgs" asciidoc libpython3.4-dev
+        # shellcheck disable=SC2086
+        apt_install $pyqt_pkgs asciidoc libpython3.4-dev
         asciidoc --version
         check_pyqt
         ;;
diff --git a/scripts/dev/ci/travis_run.sh b/scripts/dev/ci/travis_run.sh
index ebbd120246..e658ad0a2f 100644
--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -1,7 +1,7 @@
 #!/usr/bin/env bash
 
 if [[ $DOCKER ]]; then
-    docker run --privileged -v "$PWD:/outside thecompiler/qutebrowser:$DOCKER"
+    docker run --privileged -v "$PWD:/outside" "thecompiler/qutebrowser:$DOCKER"
 else
     args=()
     [[ $TESTENV == docs ]] && args=('--no-authors')
diff --git a/scripts/dev/quit_segfault_test.sh b/scripts/dev/quit_segfault_test.sh
index 90933a9650..389f125b9e 100755
--- a/scripts/dev/quit_segfault_test.sh
+++ b/scripts/dev/quit_segfault_test.sh
@@ -5,8 +5,8 @@
 echo > crash.log
 while :; do
     exit=0
-    while (( exit == 0)); do
-        duration=$(( RANDOM % 10000))
+    while (( exit == 0 )); do
+        duration=$(( RANDOM % 10000 ))
         python3 -m qutebrowser --debug ":later $duration quit" http://www.heise.de/
         exit=$?
     done
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -53,6 +53,7 @@ matrix:
       python: null
       node_js: "lts/*"
     - os: linux
+      language: generic
       env: TESTENV=shellcheck
       services: docker
   fast_finish: true

--- a/.travis.yml
+++ b/.travis.yml
@@ -52,6 +52,9 @@ matrix:
       language: node_js
       python: null
       node_js: "lts/*"
+    - os: linux
+      env: TESTENV=shellcheck
+      services: docker
   fast_finish: true
 
 cache:

--- a/misc/userscripts/cast
+++ b/misc/userscripts/cast
@@ -144,7 +144,7 @@ fi
 pkill -f "${program_}"
 
 # start youtube download in stream mode (-o -) into temporary file
-youtube-dl -qo - "$1" > ${file_to_cast} &
+youtube-dl -qo - "$1" > "${file_to_cast}" &
 ytdl_pid=$!
 
 msg info "Casting $1" >> "$QUTE_FIFO"
@@ -153,4 +153,4 @@ tail -F "${file_to_cast}" | ${program_} -
 
 # cleanup remaining background process and file on disk
 kill ${ytdl_pid}
-rm -rf ${tmpdir}
+rm -rf "${tmpdir}"

--- a/misc/userscripts/dmenu_qutebrowser
+++ b/misc/userscripts/dmenu_qutebrowser
@@ -41,7 +41,7 @@
 [ -z "$QUTE_URL" ] && QUTE_URL='http://google.com'
 
 url=$(echo "$QUTE_URL" | cat - "$QUTE_CONFIG_DIR/quickmarks" "$QUTE_DATA_DIR/history" | dmenu -l 15 -p qutebrowser)
-url=$(echo "$url" | sed -E 's/[^ ]+ +//g' | egrep "https?:" || echo "$url")
+url=$(echo "$url" | sed -E 's/[^ ]+ +//g' | grep -E "https?:" || echo "$url")
 
 [ -z "${url// }" ] && exit
 

--- a/misc/userscripts/format_json
+++ b/misc/userscripts/format_json
@@ -22,15 +22,14 @@
 # default style to monokai if none is provided
 STYLE=${1:-monokai}
 # format json using jq
-FORMATTED_JSON="$(cat "$QUTE_TEXT" | jq '.')"
-
-# if jq command failed or formatted json is empty, assume failure and terminate
-if [ $? -ne 0 ] || [ -z "$FORMATTED_JSON" ]; then
+if ! FORMATTED_JSON="$(jq . "$QUTE_TEXT")" ||  [ -z "$FORMATTED_JSON" ]; then
     echo "Invalid json, aborting..."
     exit 1
 fi
 
 # calculate the filesize of the json document
+# parsing the output of ls should be fine in this case since we set the block size
+# shellcheck disable=SC2012
 FILE_SIZE=$(ls -s --block-size=1048576 "$QUTE_TEXT" | cut -d' ' -f1)
 
 # use pygments to pretty-up the json (syntax highlight) if file is less than 10MB

--- a/misc/userscripts/format_json
+++ b/misc/userscripts/format_json
@@ -35,12 +35,12 @@ FILE_SIZE=$(ls -s --block-size=1048576 "$QUTE_TEXT" | cut -d' ' -f1)
 
 # use pygments to pretty-up the json (syntax highlight) if file is less than 10MB
 if [ "$FILE_SIZE" -lt "10" ]; then
-    FORMATTED_JSON="$(echo "$FORMATTED_JSON" | pygmentize -l json -f html -O full,style=$STYLE)"
+    FORMATTED_JSON="$(echo "$FORMATTED_JSON" | pygmentize -l json -f html -O full,style="$STYLE")"
 fi
 
 # create a temp file and write the formatted json to that file
 TEMP_FILE="$(mktemp --suffix '.html')"
-echo "$FORMATTED_JSON" > $TEMP_FILE
+echo "$FORMATTED_JSON" > "$TEMP_FILE"
 
 
 # send the command to qutebrowser to open the new file containing the formatted json

--- a/misc/userscripts/format_json
+++ b/misc/userscripts/format_json
@@ -1,4 +1,5 @@
 #!/bin/sh
+set -euo pipefail
 #
 # Behavior:
 #   Userscript for qutebrowser which will take the raw JSON text of the current
@@ -19,28 +20,23 @@
 #
 # Bryan Gilbert, 2017
 
+# do not run pygmentize on files larger than this amount of bytes
+MAX_SIZE_PRETTIFY=10485760  # 10 MB
 # default style to monokai if none is provided
 STYLE=${1:-monokai}
-# format json using jq
-if ! FORMATTED_JSON="$(jq . "$QUTE_TEXT")" ||  [ -z "$FORMATTED_JSON" ]; then
-    echo "Invalid json, aborting..."
-    exit 1
+
+TEMP_FILE="$(mktemp)"
+jq . "$QUTE_TEXT" >"$TEMP_FILE"
+
+# try GNU stat first and then OSX stat if the former fails
+FILE_SIZE=$(
+  stat --printf="%s" "$TEMP_FILE" 2>/dev/null ||
+  stat -f%z "$TEMP_FILE" 2>/dev/null
+)
+if [ "$FILE_SIZE" -lt "$MAX_SIZE_PRETTIFY" ]; then
+    pygmentize -l json -f html -O full,style="$STYLE" <"$TEMP_FILE" >"${TEMP_FILE}_"
+    mv -f "${TEMP_FILE}_" "$TEMP_FILE"
 fi
-
-# calculate the filesize of the json document
-# parsing the output of ls should be fine in this case since we set the block size
-# shellcheck disable=SC2012
-FILE_SIZE=$(ls -s --block-size=1048576 "$QUTE_TEXT" | cut -d' ' -f1)
-
-# use pygments to pretty-up the json (syntax highlight) if file is less than 10MB
-if [ "$FILE_SIZE" -lt "10" ]; then
-    FORMATTED_JSON="$(echo "$FORMATTED_JSON" | pygmentize -l json -f html -O full,style="$STYLE")"
-fi
-
-# create a temp file and write the formatted json to that file
-TEMP_FILE="$(mktemp --suffix '.html')"
-echo "$FORMATTED_JSON" > "$TEMP_FILE"
-
 
 # send the command to qutebrowser to open the new file containing the formatted json
 echo "open -t file://$TEMP_FILE" >> "$QUTE_FIFO"

--- a/misc/userscripts/open_download
+++ b/misc/userscripts/open_download
@@ -76,6 +76,7 @@ crop-first-column() {
 ls-files() {
     # add the slash at the end of the download dir enforces to follow the
     # symlink, if the DOWNLOAD_DIR itself is a symlink
+    # shellcheck disable=SC2010
     ls -Q --quoting-style escape    -h  -o -1 -A -t "${DOWNLOAD_DIR}/" \
         | grep '^[-]' \
         | cut -d' ' -f3- \
@@ -94,7 +95,7 @@ fi
 line=$(printf "%s\n" "${entries[@]}"  \
        | crop-first-column 55 \
        | column -s $'\t' -t \
-       | $ROFI_CMD "${rofi_default_args[@]}" $ROFI_ARGS) || true
+       | $ROFI_CMD "${rofi_default_args[@]}" "$ROFI_ARGS") || true
 if [ -z "$line" ];  then
     exit 0
 fi

--- a/misc/userscripts/open_download
+++ b/misc/userscripts/open_download
@@ -92,7 +92,7 @@ if [ "${#entries[@]}" -eq 0 ] ; then
     die "Download directory »${DOWNLOAD_DIR}« empty"
 fi
 
-line=$(printf "%s\n" "${entries[@]}"  \
+line=$(printf '%s\n' "${entries[@]}"  \
        | crop-first-column 55 \
        | column -s $'\t' -t \
        | $ROFI_CMD "${rofi_default_args[@]}" "$ROFI_ARGS") || true

--- a/misc/userscripts/password_fill
+++ b/misc/userscripts/password_fill
@@ -236,7 +236,7 @@ pass_backend() {
         if ((match_line)) ; then
             # add entries with matching URL-tag
             while read -r -d "" passfile ; do
-                if $GPG "${GPG_OPTS}" -d "$passfile" \
+                if $GPG "${GPG_OPTS[@]}" -d "$passfile" \
                      | grep --max-count=1 -iE "${match_line_pattern}${url}" > /dev/null
                 then
                     passfile="${passfile#$PREFIX}"
@@ -269,7 +269,7 @@ pass_backend() {
                     break
                 fi
             fi
-        done < <($GPG "${GPG_OPTS}" -d "$path" )
+        done < <($GPG "${GPG_OPTS[@]}" -d "$path" )
     }
 }
 # =======================================================
@@ -283,7 +283,7 @@ secret_backend() {
     query_entries() {
         local domain="$1"
         while read -r line ; do
-            if [[ "$line" =~ "attribute.username = " ]] ; then
+            if [[ "$line" == "attribute.username ="* ]] ; then
                 files+=("$domain ${line#${BASH_REMATCH[0]}}")
             fi
         done < <( secret-tool search --unlock --all domain "$domain" 2>&1 )

--- a/misc/userscripts/password_fill
+++ b/misc/userscripts/password_fill
@@ -64,7 +64,7 @@ die() {
 javascript_escape() {
     # print the first argument in an escaped way, such that it can safely
     # be used within javascripts double quotes
-    sed "s,[\\\'\"],\\\&,g" <<< "$1"
+    sed "s,[\\\\'\"],\\\\&,g" <<< "$1"
 }
 
 # ======================================================= #

--- a/misc/userscripts/password_fill
+++ b/misc/userscripts/password_fill
@@ -178,7 +178,7 @@ choose_entry_menu() {
     if [ "$nr" -eq 1 ] && ! ((menu_if_one_entry)) ; then
         file="${files[0]}"
     else
-        file=$( printf "%s\n" "${files[@]}" | "${MENU_COMMAND[@]}" )
+        file=$( printf '%s\n' "${files[@]}" | "${MENU_COMMAND[@]}" )
     fi
 }
 
@@ -303,6 +303,7 @@ pass_backend
 QUTE_CONFIG_DIR=${QUTE_CONFIG_DIR:-${XDG_CONFIG_HOME:-$HOME/.config}/qutebrowser/}
 PWFILL_CONFIG=${PWFILL_CONFIG:-${QUTE_CONFIG_DIR}/password_fill_rc}
 if [ -f "$PWFILL_CONFIG" ] ; then
+    # shellcheck source=/dev/null
     source "$PWFILL_CONFIG"
 fi
 init
@@ -311,7 +312,7 @@ simplify_url "$QUTE_URL"
 query_entries "${simple_url}"
 no_entries_found
 # remove duplicates
-mapfile -t files < <(printf "%s\n" "${files[@]}" | sort | uniq )
+mapfile -t files < <(printf '%s\n' "${files[@]}" | sort | uniq )
 choose_entry
 if [ -z "$file" ] ; then
     # choose_entry didn't want any of these entries

--- a/misc/userscripts/qutedmenu
+++ b/misc/userscripts/qutedmenu
@@ -35,17 +35,11 @@ get_selection() {
 
 # Main
 # https://github.com/halfwit/dotfiles/blob/master/.config/dmenu/font
-if [[ -s $confdir/dmenu/font ]]; then
-    read -r font < "$confdir"/dmenu/font
-fi
+[[ -s $confdir/dmenu/font ]] && read -r font < "$confdir"/dmenu/font
 
-if [[ $font ]]; then
-    opts+=(-fn "$font")
-fi
+[[ $font ]] && opts+=(-fn "$font")
 
-if [[ -s $optsfile ]]; then
-    source "$optsfile"
-fi
+[[ -s $optsfile ]] && source "$optsfile"
 
 url=$(get_selection)
 url=${url/*http/http}

--- a/misc/userscripts/qutedmenu
+++ b/misc/userscripts/qutedmenu
@@ -39,6 +39,7 @@ get_selection() {
 
 [[ $font ]] && opts+=(-fn "$font")
 
+# shellcheck source=/dev/null
 [[ -s $optsfile ]] && source "$optsfile"
 
 url=$(get_selection)

--- a/misc/userscripts/rss
+++ b/misc/userscripts/rss
@@ -57,7 +57,7 @@ notice () {
 
 # Update a database of a feed and open new URLs
 read_items () {
-    cd read_urls || return
+    cd read_urls || return 1
     feed_file="$(echo "$1" | tr -d /)"
     feed_temp_file="$(mktemp "$feed_file.tmp.XXXXXXXXXX")"
     feed_new_items="$(mktemp "$feed_file.new.XXXXXXXXXX")"
@@ -85,7 +85,7 @@ if [ ! -d "$config_dir/read_urls" ]; then
     mkdir -p "$config_dir/read_urls"
 fi
 
-cd "$config_dir" || exit
+cd "$config_dir" || exit 1
 
 if [ $# != 0 ]; then
     for arg in "$@"; do

--- a/misc/userscripts/rss
+++ b/misc/userscripts/rss
@@ -32,7 +32,7 @@ add_feed () {
     if grep -Fq "$1" "feeds"; then
         notice "$1 is saved already."
     else
-        printf "%s\n" "$1" >> "feeds"
+        printf '%s\n' "$1" >> "feeds"
     fi
 }
 
@@ -57,7 +57,7 @@ notice () {
 
 # Update a database of a feed and open new URLs
 read_items () {
-    cd read_urls
+    cd read_urls || return
     feed_file="$(echo "$1" | tr -d /)"
     feed_temp_file="$(mktemp "$feed_file.tmp.XXXXXXXXXX")"
     feed_new_items="$(mktemp "$feed_file.new.XXXXXXXXXX")"
@@ -75,7 +75,7 @@ read_items () {
         cat "$feed_new_items" >> "$feed_file"
         sort -o "$feed_file" "$feed_file"
         rm "$feed_temp_file" "$feed_new_items"
-    fi | while read item; do
+    fi | while read -r item; do
         echo "open -t $item" > "$QUTE_FIFO"
     done
 }
@@ -85,7 +85,7 @@ if [ ! -d "$config_dir/read_urls" ]; then
     mkdir -p "$config_dir/read_urls"
 fi
 
-cd "$config_dir"
+cd "$config_dir" || exit
 
 if [ $# != 0 ]; then
     for arg in "$@"; do
@@ -115,7 +115,7 @@ if < /dev/null grep --help 2>&1 | grep -q -- -a; then
     text_only="-a"
 fi
 
-while read feed_url; do
+while read -r feed_url; do
     read_items "$feed_url" &
 done < "$config_dir/feeds"
 

--- a/misc/userscripts/taskadd
+++ b/misc/userscripts/taskadd
@@ -25,12 +25,12 @@
 [[ $QUTE_MODE == 'hints' ]] && title=$QUTE_SELECTED_TEXT || title=$QUTE_TITLE
 
 # try to add the task and grab the output
-msg="$(task add $title $@ 2>&1)"
+msg="$(task add "$title" "$*" 2>&1)"
 
 if [[ $? == 0 ]]; then
     # annotate the new task with the url, send the output back to the browser
     task +LATEST annotate "$QUTE_URL"
-    echo "message-info '$msg'" >> $QUTE_FIFO
+    echo "message-info '$msg'" >> "$QUTE_FIFO"
 else
-    echo "message-error '$msg'" >> $QUTE_FIFO
+    echo "message-error '$msg'" >> "$QUTE_FIFO"
 fi

--- a/misc/userscripts/taskadd
+++ b/misc/userscripts/taskadd
@@ -25,9 +25,7 @@
 [[ $QUTE_MODE == 'hints' ]] && title=$QUTE_SELECTED_TEXT || title=$QUTE_TITLE
 
 # try to add the task and grab the output
-msg="$(task add "$title" "$*" 2>&1)"
-
-if [[ $? == 0 ]]; then
+if msg="$(task add "$title" "$*" 2>&1)"; then
     # annotate the new task with the url, send the output back to the browser
     task +LATEST annotate "$QUTE_URL"
     echo "message-info '$msg'" >> "$QUTE_FIFO"

--- a/misc/userscripts/view_in_mpv
+++ b/misc/userscripts/view_in_mpv
@@ -50,7 +50,7 @@ msg() {
 MPV_COMMAND=${MPV_COMMAND:-mpv}
 # Warning: spaces in single flags are not supported
 MPV_FLAGS=${MPV_FLAGS:- --force-window --no-terminal --keep-open=yes --ytdl --ytdl-raw-options=yes-playlist=}
-video_command=( "$MPV_COMMAND"  $MPV_FLAGS )
+IFS=" " read -r -a video_command <<< "$MPV_COMMAND $MPV_FLAGS"
 
 js() {
 cat <<EOF

--- a/scripts/dev/ci/travis_backtrace.sh
+++ b/scripts/dev/ci/travis_backtrace.sh
@@ -6,7 +6,7 @@
 
 case $TESTENV in
     py3*-pyqt*)
-        exe=$(readlink -f .tox/$TESTENV/bin/python)
+        exe=$(readlink -f ".tox/$TESTENV/bin/python")
         full=
         ;;
     *)

--- a/scripts/dev/ci/travis_backtrace.sh
+++ b/scripts/dev/ci/travis_backtrace.sh
@@ -15,4 +15,4 @@ case $TESTENV in
         ;;
 esac
 
-find . -name *.core -o -name core -exec gdb --batch --quiet -ex "thread apply all bt $full" "$exe" {} \;
+find . \( -name "*.core" -o -name core \) -exec gdb --batch --quiet -ex "thread apply all bt $full" "$exe" {} \;

--- a/scripts/dev/ci/travis_install.sh
+++ b/scripts/dev/ci/travis_install.sh
@@ -96,6 +96,8 @@ case $TESTENV in
     eslint)
         npm_install eslint
         ;;
+    shellcheck)
+        ;;
     *)
         pip_install pip
         pip_install -r misc/requirements/requirements-tox.txt

--- a/scripts/dev/ci/travis_install.sh
+++ b/scripts/dev/ci/travis_install.sh
@@ -27,17 +27,17 @@ travis_retry() {
     local count=1
     while (( count < 3 )); do
         if (( result != 0 )); then
-            echo -e "\n${ANSI_RED}The command \"$@\" failed. Retrying, $count of 3.${ANSI_RESET}\n" >&2
+            echo -e "\n${ANSI_RED}The command \"$*\" failed. Retrying, $count of 3.${ANSI_RESET}\n" >&2
         fi
         "$@"
         result=$?
         (( result == 0 )) && break
-        count=$(($count + 1))
+        count=$(( count + 1 ))
         sleep 1
     done
 
     if (( count > 3 )); then
-        echo -e "\n${ANSI_RED}The command \"$@\" failed 3 times.${ANSI_RESET}\n" >&2
+        echo -e "\n${ANSI_RED}The command \"$*\" failed 3 times.${ANSI_RESET}\n" >&2
     fi
 
     return $result

--- a/scripts/dev/ci/travis_install.sh
+++ b/scripts/dev/ci/travis_install.sh
@@ -21,13 +21,13 @@
 # Stolen from https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/templates/header.sh
 # and adjusted to use ((...))
 travis_retry() {
-    local ANSI_RED="\033[31;1m"
-    local ANSI_RESET="\033[0m"
+    local ANSI_RED='\033[31;1m'
+    local ANSI_RESET='\033[0m'
     local result=0
     local count=1
     while (( count < 3 )); do
         if (( result != 0 )); then
-            echo -e "\n${ANSI_RED}The command \"$*\" failed. Retrying, $count of 3.${ANSI_RESET}\n" >&2
+            echo -e "\\n${ANSI_RED}The command \"$*\" failed. Retrying, $count of 3.${ANSI_RESET}\\n" >&2
         fi
         "$@"
         result=$?
@@ -37,7 +37,7 @@ travis_retry() {
     done
 
     if (( count > 3 )); then
-        echo -e "\n${ANSI_RED}The command \"$*\" failed 3 times.${ANSI_RESET}\n" >&2
+        echo -e "\\n${ANSI_RED}The command \"$*\" failed 3 times.${ANSI_RESET}\\n" >&2
     fi
 
     return $result

--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -16,14 +16,12 @@ elif [[ $TESTENV == eslint ]]; then
     eslint --color --report-unused-disable-directives .
 elif [[ $TESTENV == shellcheck ]]; then
     dev_scripts=$( find scripts/dev/ -name '*.sh' -print0 | xargs -0 )
-    # false positive: we are using 'find -exec +'
-    # shellcheck disable=SC2038
     userscripts=$( find misc/userscripts/ -type f -exec grep -lE '[/ ][bd]ash$|[/ ]sh$|[/ ]ksh$' {} + | xargs )
     IFS=" " read -r -a scripts <<< "$dev_scripts $userscripts"
     docker run \
            -v "$PWD:/outside" \
 	   -w /outside \
-           koalaman/shellcheck:stable "${scripts[@]}"
+           koalaman/shellcheck:latest "${scripts[@]}"
 else
     args=()
     [[ $TRAVIS_OS_NAME == osx ]] && args=('--qute-bdd-webengine' '--no-xvfb')

--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -14,6 +14,16 @@ elif [[ $TESTENV == eslint ]]; then
     # travis env
     cd qutebrowser/javascript || exit 1
     eslint --color --report-unused-disable-directives .
+elif [[ $TESTENV == shellcheck ]]; then
+    dev_scripts=$( find scripts/dev/ -name '*.sh' -print0 | xargs -0 )
+    # false positive: we are using 'find -exec +'
+    # shellcheck disable=SC2038
+    userscripts=$( find misc/userscripts/ -type f -exec grep -lE '[/ ][bd]ash$|[/ ]sh$|[/ ]ksh$' {} + | xargs )
+    IFS=" " read -r -a scripts <<< "$dev_scripts $userscripts"
+    docker run \
+           -v "$PWD:/outside" \
+	   -w /outside \
+           koalaman/shellcheck:stable "${scripts[@]}"
 else
     args=()
     [[ $TRAVIS_OS_NAME == osx ]] && args=('--qute-bdd-webengine' '--no-xvfb')

--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -16,6 +16,8 @@ elif [[ $TESTENV == eslint ]]; then
     eslint --color --report-unused-disable-directives .
 elif [[ $TESTENV == shellcheck ]]; then
     dev_scripts=$( find scripts/dev/ -name '*.sh' -print0 | xargs -0 )
+    # false positive: we are using 'find -exec +'
+    # shellcheck disable=SC2038
     userscripts=$( find misc/userscripts/ -type f -exec grep -lE '[/ ][bd]ash$|[/ ]sh$|[/ ]ksh$' {} + | xargs )
     IFS=" " read -r -a scripts <<< "$dev_scripts $userscripts"
     docker run \

--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -22,7 +22,7 @@ elif [[ $TESTENV == shellcheck ]]; then
     IFS=" " read -r -a scripts <<< "$dev_scripts $userscripts"
     docker run \
            -v "$PWD:/outside" \
-	   -w /outside \
+           -w /outside \
            koalaman/shellcheck:latest "${scripts[@]}"
 else
     args=()

--- a/scripts/dev/ci/travis_run.sh
+++ b/scripts/dev/ci/travis_run.sh
@@ -15,11 +15,11 @@ elif [[ $TESTENV == eslint ]]; then
     cd qutebrowser/javascript || exit 1
     eslint --color --report-unused-disable-directives .
 elif [[ $TESTENV == shellcheck ]]; then
-    dev_scripts=$( find scripts/dev/ -name '*.sh' -print0 | xargs -0 )
-    # false positive: we are using 'find -exec +'
-    # shellcheck disable=SC2038
-    userscripts=$( find misc/userscripts/ -type f -exec grep -lE '[/ ][bd]ash$|[/ ]sh$|[/ ]ksh$' {} + | xargs )
-    IFS=" " read -r -a scripts <<< "$dev_scripts $userscripts"
+    SCRIPTS=$( mktemp )
+    find scripts/dev/ -name '*.sh' >"$SCRIPTS"
+    find misc/userscripts/ -type f -exec grep -lE '[/ ][bd]ash$|[/ ]sh$|[/ ]ksh$' {} + >>"$SCRIPTS"
+    mapfile -t scripts <"$SCRIPTS"
+    rm -f "$SCRIPTS"
     docker run \
            -v "$PWD:/outside" \
            -w /outside \

--- a/scripts/dev/download_release.sh
+++ b/scripts/dev/download_release.sh
@@ -11,7 +11,7 @@ if [[ $# != 1 ]]; then
     exit 1
 fi
 
-cd "$tmpdir"
+cd "$tmpdir" || exit
 mkdir windows
 
 base="https://github.com/qutebrowser/qutebrowser/releases/download/v$1"
@@ -21,13 +21,13 @@ wget "$base/qutebrowser-$1.tar.gz.asc" || exit 1
 wget "$base/qutebrowser-$1.dmg" || exit 1
 wget "$base/qutebrowser_${1}-1_all.deb" || exit 1
 
-cd windows
+cd windows || exit
 wget "$base/qutebrowser-${1}-amd64.msi" || exit 1
 wget "$base/qutebrowser-${1}-win32.msi" || exit 1
 wget "$base/qutebrowser-${1}-windows-standalone-amd64.zip" || exit 1
 wget "$base/qutebrowser-${1}-windows-standalone-win32.zip" || exit 1
 
 dest="/srv/http/qutebrowser/releases/v$1"
-cd "$oldpwd"
+cd "$oldpwd" || exit
 sudo mv "$tmpdir" "$dest"
 sudo chown -R http:http "$dest"

--- a/scripts/dev/download_release.sh
+++ b/scripts/dev/download_release.sh
@@ -1,4 +1,5 @@
 #!/bin/bash
+set -e
 
 # This script downloads the given release from GitHub so we can mirror it on
 # qutebrowser.org.
@@ -11,23 +12,23 @@ if [[ $# != 1 ]]; then
     exit 1
 fi
 
-cd "$tmpdir" || exit
+cd "$tmpdir"
 mkdir windows
 
 base="https://github.com/qutebrowser/qutebrowser/releases/download/v$1"
 
-wget "$base/qutebrowser-$1.tar.gz" || exit 1
-wget "$base/qutebrowser-$1.tar.gz.asc" || exit 1
-wget "$base/qutebrowser-$1.dmg" || exit 1
-wget "$base/qutebrowser_${1}-1_all.deb" || exit 1
+wget "$base/qutebrowser-$1.tar.gz"
+wget "$base/qutebrowser-$1.tar.gz.asc"
+wget "$base/qutebrowser-$1.dmg"
+wget "$base/qutebrowser_${1}-1_all.deb"
 
-cd windows || exit
-wget "$base/qutebrowser-${1}-amd64.msi" || exit 1
-wget "$base/qutebrowser-${1}-win32.msi" || exit 1
-wget "$base/qutebrowser-${1}-windows-standalone-amd64.zip" || exit 1
-wget "$base/qutebrowser-${1}-windows-standalone-win32.zip" || exit 1
+cd windows
+wget "$base/qutebrowser-${1}-amd64.msi"
+wget "$base/qutebrowser-${1}-win32.msi"
+wget "$base/qutebrowser-${1}-windows-standalone-amd64.zip"
+wget "$base/qutebrowser-${1}-windows-standalone-win32.zip"
 
 dest="/srv/http/qutebrowser/releases/v$1"
-cd "$oldpwd" || exit
+cd "$oldpwd"
 sudo mv "$tmpdir" "$dest"
 sudo chown -R http:http "$dest"

--- a/scripts/dev/quit_segfault_test.sh
+++ b/scripts/dev/quit_segfault_test.sh
@@ -1,14 +1,12 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
-if [[ $PWD == */scripts ]]; then
-    cd ..
-fi
+[[ $PWD == */scripts ]] && cd ..
 
 echo > crash.log
 while :; do
     exit=0
-    while (( $exit == 0)); do
-        duration=$(($RANDOM%10000))
+    while (( exit == 0 )); do
+        duration=$(( RANDOM % 10000 ))
         python3 -m qutebrowser --debug ":later $duration quit" http://www.heise.de/
         exit=$?
     done

